### ISSUE TITLE
BlockRepository Cache

### ISSUE THREAD

**Issue by noescape00** [stratisproject/StratisBitcoinFullNode#1030]
## Problem

Right now we need to have the block saved on the hard drive before we are to perform some actions. For example: we won't start announcing the blocks that we've just downloaded or mined until the moment they are persisted to the disk. 

This causes an unnecessary slow down of several processes that node does, most importantly: blocks propagation. 

## Solution
Implementing a cache for the block repository will fix this issue- we will be able to save blocks to the repository instantly and retrieve blocks as soon as they are there. Meanwhile blocks will be saved to the hard drive in background.

This can be done in 2 ways: 

1. Building something like `BlockRepositoryWithCache` on top of the `BlockRepository` that will have an internal blocks list that will operate as cache and that will persist items from that list in the background.
2. Utilizing `BlockStoreCache` class by saving blocks there on every BlockRepository.PutAsync call before we start saving them to the hard drive. Also we will need to use `BlockStoreCache` in every situation when we need access to the blocks that are potentially not persisted on disk yet. 

Personally I'd go with first option because it's more obvious\logical one. Plus it will allow getting rid of  `BlockStoreCache` since it's very usage is to avoid getting data from the hard drive all the time, and we will be able to do that with `BlockRepositoryWithCache`. Though 1st option will require a bit more work to do comparing to just caching the block every time before we save it to the hard drive.

**Comment by Aprogiena** [comment]
if there exist such a thing, why not use it? 

can we identify all components that access block store directly? are there any such components that should rather go through cache?

what are the current rules on adding a block to the cache? when are currently cached blocks removed from cache? 

for the purpose of announcement, we want to make sure the cache always contains last X blocks

**Comment by noescape00** [comment]
>what are the current rules on adding a block to the cache? when are currently cached blocks removed from cache?

Every time we do `IBlockStoreCache.GetBlockAsync` it will first try to get the block from the cache. If it's not there it will get the block from the BlockRepository and add it to the cache (the block will be deleted from it in 10 minutes). Btw this cache behavior is also kind of wrong because in IBD we are ending up with 1gb worth of RAM cached blocks (but that's something for another issue or PR).

Other then that we don't add block to the cache (thow there is an interface for that but we don't use it).

**Comment by noescape00** [comment]
>can we identify all components that access block store directly?

We can.

>are there any such components that should rather go through cache?

I think if all components go through cache in order to get a block by id we will only benefit from it.

**Comment by noescape00** [comment]
>for the purpose of announcement, we want to make sure the cache always contains last X blocks

Yup. So we can manually do IBlockStoreCache.AddToCache on every IBlockRepository.PutAsync call or just merge those 2 classes together so we don't need to decide what to use- cache or block repository, it will be decided by the inner logic.

**Comment by Aprogiena** [comment]
> Other then that we don't add block to the cache (thow there is an interface for that but we don't use it).

OK, so from your description it seems that the current rules are far from optimal. For example, having a block being removed from the cache after 10 minutes is not very helpful on Bitcoin network. I agree that a size based limit would make better sense. Maybe each block should have last access time with it in the cache and if the memory limit was reached, we would remove say 25 % of the least recently used entries. Something like that. Time limit does not make much sense to me here.

Then you say there is an interface for adding new blocks from different areas - I think this is what we want to use. When we mine a block or receive it from the puller, we should add it to the cache. Probably part of accept block routine in consensus loop.


> I think if all components go through the cache in order to get a block we will only benefit from it.

I don't think this is correct. For example, the very BlockStoreSignaled does not. So my question was about identifying all such components.

> Yup. So we can manually do IBlockStoreCache.AddToCache on every IBlockRepository.PutAsync call or just merge those 2 classes together (make a cache-supported implementation on top of the BlockRepository) so we don't need to decide what to use: cache or block repository- it will be decided by the inner logic.

My previous questions were asked to identify which components use what and then discuss why is that. For example, it seems mistaken that BlockStoreSignaled does not use cache. 

I think very specific components may need not using cache, but most of the components should use the cache and not to access block store directly. So it seems beneficial to me, to make this analysis and then refactor those components to use the cache and also to improve the cache itself. 

Merging block store with cache - I'm not sure it is needed. I can't see enough advantage in doing that to justify extra effort. I think using the cache intentionally or the store directly would be fine for all use cases.

**Comment by noescape00** [comment]
>OK, so from your description it seems that the current rules are far from optimal. For example, having a block being removed from the cache after 10 minutes is not very helpful on Bitcoin network. I agree that a size based limit would make better sense. Maybe each block should have last access time with it in the cache and if the memory limit was reached, we would remove say 25 % of the least recently used entries. Something like that.

Hopefully we dont use cache too much atm so it hardly causes any trouble. But if we are to utilize ir or merge with BR we will need to rethink it's logic. 

>Time limit does not make much sense to me here.

I think it's a good idea to keep both of the limits- time & capacity. At least that's what they usually do in solutions like Redis. 

>Most of them yes, maybe not all of them. The question is - which components need to be changed?

Hm, it will take some time to figure out what components really need this. Right now I'm only sure about components that do the blocks propagation so we can start by doing just that.

**Comment by noescape00** [comment]
>So it seems beneficial to me, to make this analysis and then refactor those components to use the cache and also to improve the cache itself.
>Merging block store with cache - I'm not sure it is needed. 

Agreed. Thanks for the feedback. 

I'll start working on improving the cache and integrating it into the components one by one in separate PRs, starting from block announcing one.

**Comment by Aprogiena** [comment]
I would probably wait for input from @dangershony about this, because just the two of us, we may have incomplete information.

**Comment by noescape00** [comment]
@Aprogiena good point. Will do some research or work on smth else in the meantime.

**Comment by Aprogiena** [comment]
> I think it's a good idea to keep both of the limits- time & capacity. At least that's what they usually do in solutions like Redis.

I mean, if you have say 50 MB reserved for the cache. You fill the cache to say 40 MB, still 10 MB is empty. Why would you removed something from the cache, just because it was added more than 1 hour ago? You don't really save any memory because you do count with those 50 MB for the cache. And the only thing you will achieve is slowing down the performance if the block you removed is needed again. Removing this data from cache seems to be a cheap operation, so I'm not sure it is worth doing that in advance (before it is needed).

**Comment by noescape00** [comment]
>You don't really save any memory because you do count with those 50 MB for the cache.

It's not like cache has that 50mb worth of ram always allocated for it. If we free the cache it will actually garbage collect released items and will save some RAM.

**Comment by Aprogiena** [comment]
Yes, but I don't see why that would be useful. Cache is used to trade this memory resource for speedup.

**Comment by noescape00** [comment]
>Yes, but I don't see why that would be useful. 

I guess if we run the node on a stanalone PC with 32GB of RAM it's unnecessary, but for sensitive cases like running the node on raspberry pi or smth like it this might be useful.

**Comment by noescape00** [comment]
Also take in count that implementing this comes cheaply. It's a prat of the MemoryCache interface already, just need to configure it properly.

**Comment by Aprogiena** [comment]
It would be more useful to be able to define the upper limit of RAM for the cache, because otherwise you will always need to count with those extra 50 MB and have it there ready for the cache. So in order to support such low end setups, you need configuration option. The removal from the cache won't help.

**Comment by noescape00** [comment]
> So in order to support such low end setups, you need configuration option. The removal from the cache won't help.

By configuring the upper bound for the cache size we basically telling MemoryCache to remove old items if we exceed this size. 

So yeah, that's what I meant- we set the max size, we set the expire time and that way we will always have reasonable size cache suitable for all setups.

**Comment by Aprogiena** [comment]
What i meant was to expose this configuration to the user, so that if someone wants to run it on low memory machine, they can tweak it without recompilation.

**Comment by noescape00** [comment]
@Aprogiena yes, making max cache size configurable is a good idea.

**Comment by noescape00** [comment]
@dangershony what do you think about this?

**Comment by dangershony** [comment]
I agree on a configurable max size, regarding time limiting initially I would say to add a limit of 1 hour but @Aprogiena has good points if we already assume this memory to be allocated then we just leave the blocks in cache otherwise I see no benefit on removing them, it does mean how ever the memory will never clear until new blocks replace old blocks after a sync.

Most of the block requests probably happen from nodes that are catching up (not IBD) and so its useful to keep the last blocks in memory.

I would say we need a PR that will do 
- Call the method `IBlockStoreCache.AddToCache()` before we broadcast to a peer 
- Remove the time limit (or push it to be a much bigger number)
- Set a max cache size and make it configurable (we need to make sure that the older cache entries expire first)

**Comment by dangershony** [comment]
One more consequence is invalid blocks in cache if a reorg happened (not a big deal)

**Comment by Aprogiena** [comment]
such blocks are not invalid, but i get your point

such blocks would occupy the cache unless we remove them, but i think we can remove them during unwind, so no issue i guess

i also think it is reasonable to have different default value for the cache size for different networks

**Comment by dangershony** [comment]
Yeah sorry I meant orphan blocks.
Different size networks indeed are a consideration but in theory its the size of all trx in the cached blocks is that matter, with stratis it means more cache entries while bitcoin its bigger cache entries.

**Comment by Aprogiena** [comment]
yes, but it makes sense to have something like 300 MB for bitcoin as default 
and something like 30 MB for stratis

**Comment by dangershony** [comment]
It makes sense for now (blocks are smaller) but its a moving variable and its a bit more of a challenge to have different network values so we need to pick them carefully.

**Comment by Aprogiena** [comment]
hmm that's right, so maybe it does not make that much sense to have different values to be really future proof

**Comment by noescape00** [comment]
Maybe make cache size = (avg block size based on last 100 blocks) * 300? 
That way we won't have to address that in the future and change cache value every time the block size doubles.

**Comment by Aprogiena** [comment]
you probably want that number to be known to the user, so that they can easily change it and know from which value they changed it to which new value

**Comment by noescape00** [comment]
@Aprogiena hm, reasonable. 
So we can go with the hardcoded value or optionally log it on startup: "BlockStore Cache initialized with X mb max size".

**Comment by noescape00** [comment]
>we need to make sure that the older cache entries expire first

Apparently Microsoft.Extensions.Caching.Memory.MemoryCache doesn't support that =(

So I'll probably need to completely refactore MemoryCache.

**Comment by Aprogiena** [comment]
are you sure MemoryCache does not use LRU?

**Comment by Aprogiena** [comment]
but even if that cache was not good enough, you don't probably need to do anything with BSC, because you'd just implement IMemoryCache

**Comment by noescape00** [comment]
Snippet:

```
static void Main(string[] args)
        {
            var memoryCacheOptions = new MemoryCacheOptions()
            {
                SizeLimit = 5
            };

            MemoryCache cache = new MemoryCache(memoryCacheOptions);

            for (int i = 0; i < 50; ++i)
            {
                string data = i + "__SOMEDATA";
                cache.Set(i, data, new MemoryCacheEntryOptions() { Size = 1, AbsoluteExpirationRelativeToNow = TimeSpan.FromMinutes(1) });
                Thread.Sleep(10);
            }

            for (int i = 0; i < 50; ++i)
            {
                if (cache.TryGetValue(i, out string data))
                {
                    Console.WriteLine(data);
                }
            }
            Console.ReadLine();
        }
```

Result:

```
41__SOMEDATA
43__SOMEDATA
45__SOMEDATA
47__SOMEDATA
49__SOMEDATA
```

or it will be 1,2,3,4,5 if you remove that Thread.Sleep(10);

**Comment by noescape00** [comment]
>but even if that cache was not good enough, you don't probably need to do anything with BSC, because you'd just implement IMemoryCache

Right now I'm thinking more about adopting some other solution (maybe [this one](http://cachemanager.michaco.net/)) and making a `IBlockStoreCache` wrapper.

**Comment by Aprogiena** [comment]
seems like an overkill, what we want is super simple to even code ourselves

**Comment by noescape00** [comment]
well alternatevely I can add a list <id,blockId> and manually delete the items instead of giving this to the inner implementation if MemoryCache. 
Sounds like a cheap enough solution to me plus we won't have to completely changing anything.

**Comment by Aprogiena** [comment]
I'd implement IMemoryCache and you'd need a linked list, not list, to be able to implement LRU
so linked list + dictionary.

**Comment by noescape00** [comment]
What the point of implementing a whole cache class if we can just tweak cleaning up mechanics of existing solution in ~20 lines?

**Comment by Aprogiena** [comment]
so you mean to still use MC just to add LRU?

**Comment by noescape00** [comment]
>so you mean to still use MC just to add LRU? 

Exactly.

**Comment by dangershony** [comment]
https://stackoverflow.com/questions/9653696/default-memory-cache-with-lru-policy

**Comment by dangershony** [comment]
By the way @noescape00  I think the example you made actually proves this does work no?
The early items have been removed (expired)

**Comment by noescape00** [comment]
The order isn't exactly correct (I'd expect 49,48,47,46,45 to be on the list) plus it acts weird when you add many items at once. 

Well, adding many items is probably not what we will do with it. 
So if we're ok with slightly weird order so it's not exactly LRU but close enough then we may keep everything as it is. 
@dangershony what do you say?

**Comment by noescape00** [comment]
Another thing about the MemoryCache is that it won't calculate used RAM size itself. Instead it allows specifying size in MemoryCacheEntryOptions on adding to the cache. So we need to do 
`block.GetSerializedSize()` which is basically doing serialization on every cache insert which we probably would like to avoid. 
Other option is to set size limits in blocks (max 300 blocks, let's say) and set size of 1 for an entry.

**Comment by Aprogiena** [comment]
@dangershony i've seen that SO, but it seems that THAT MC is different than .net core MC (at least it is different namespace), so maybe it is something else and if @noescape00 example works as he claims, then it is not LRU, although i can recheck that

@noescape00 i think we could very well have this information without calling GSS

**Comment by noescape00** [comment]
>THAT MC is different than .net core MC

It is different, right. Also there is no overload for NameValueCollection therefore we can't set 'CacheMemoryLimitMegabytes' and stuff like that

**Comment by noescape00** [comment]
>@noescape00 i think we could very well have this information without calling GSS

I guess we will have to use serialization one way or another in order to get exact size. 

https://social.msdn.microsoft.com/Forums/vstudio/en-US/96747ab7-7d89-4846-9e83-46f71b8ccc66/how-to-determine-size-of-the-c-object?forum=clr

So I suggest we use max blocks count instead of max ram .

**Comment by Aprogiena** [comment]
when we download a block, we have it in our hands, we don't need to serialize

the only problem is when we mine a block, we would like to avoid double serialization there
but it seems it is hard to avoid that in this case

**Comment by noescape00** [comment]
>when we download a block, we have it in our hands, we don't need to serialize

right now we don't have access to the size without serialization in places where bscache is currently used. So particularly in case of storing just downloaded block- yes, here we can avoid serialization because we know the size. 

But overall remember why we wanted to use the cache? It should have our block right away so we don't wait those 80-120ms before it is saved on drive. So taking that in count I think it will be reasonable to avoid any heavy calculations like serialization at all times.

**Comment by noescape00** [comment]
On the other side what's wrong with setting the limit in blocks count? Sounds way more straightforward to me.

**Comment by Aprogiena** [comment]
of course, we will avoid that, but we can still use the size

**Comment by Aprogiena** [comment]
the limit in blocks count is not that good for the user
if they have 2 GB memory, what number they should set this too?

**Comment by Aprogiena** [comment]
for our own mined blocks, i think it is ok to cheat and just estimate its size

**Comment by noescape00** [comment]
>if they have 2 GB memory, what number they should set this too?

Ok, that will make user struggle (though if user is to set this param he is probably know what he's doing). 
What about the assumption thing? 

On start- get the last block in the repo and calculate the size- assume next 1000 blocks will be of the same size.

**Comment by Aprogiena** [comment]
that wont be accurate at all, we can either use max size, or estimate base on transaction count

**Comment by noescape00** [comment]
>that wont be accurate at all

94% accurate if we calculate avg size using 100 blocks and recalculate it every 10k blocks. 

97.3% accurate if we use 1000 blocks for avg size calculation.

**Comment by noescape00** [comment]
I guess if we calculate avg tx size basing on 100 blocks and will estimate block's size basing on the tx count it has we will get results close to 100% accuracy.

**Comment by Aprogiena** [comment]
maybe if we just use previous block or even max block size we could be fine, consider how many block one mines, even with 4 % stake, you have so few blocks that almost everything is overkill, i'd probably go with previous block size for minted blocks

**Comment by Aprogiena** [comment]
> 94% accurate if we calculate avg size using 100 blocks and recalculate it every 10k blocks.
> 97.3% accurate if we use 1000 blocks for avg size calculation.
now try to run it on the whole history of bitcoin blockchain ;)

**Comment by noescape00** [comment]
>now try to run it on the whole history of bitcoin blockchain ;)

It's not like we will be using cache in IBD so why would we care about old inconsistent history?  

>maybe if we just use previous block or even max block size we could be fine, consider how many block one mines, even with 4 % stake, you have so few blocks that almost everything is overkill, i'd probably go with previous block size for minted blocks

Yup, for the mined ones it sounds fine, but there are also ones that we get from block repo first and add to cache. 
I really want to avoid any serialization here too because this might be a major delay.

**Comment by Aprogiena** [comment]
> It's not like we will be using cache in IBD so why would we care about old inconsistent history?

because i think your numbers are unreal, you should back 
test on chain history of serious size to find the real accuracy
i don't believe that any kind of avg would stay for 1000 blocks, that's 7 days in bitcoin; 10k blocks is 70 days

if you check this chart
https://blockchain.info/charts/mempool-size?timespan=1year
it does not seem like something calculated would stay for 7-70 days 

i just don't believe those numbers

**Comment by Aprogiena** [comment]
> there are also ones that we get from block repo first and add to cache.

ok, this is interesting

this leads me to an idea, we could say that the size is unknown, if it is and the first time we get it, we fill it, this should work fine for both cases - blocks mined and loaded from repo

**Comment by noescape00** [comment]
>it does not seem like something calculated would stay for 7-70 days

Hm. You got my the idea wrong: 

When we start the node we initialize the cache without a limit and start a setLimit task
This task from now on will be launched every 2000 blocks. 

This task would get top 100 blocks in our block repository and serialize them, summ the sizes and devide by 100 = average block size. Basing on average block size we now can estimate when cache is our of the limit. 

So it's not supposed to stay for 7-70 days, just for 1

**Comment by noescape00** [comment]
>this leads me to an idea, we could say that the size is unknown, if it is and the first time we get it, we fill it, this should work fine for both cases - blocks mined and loaded from repo

Yup, that's what I'm talking about =)

**Comment by Aprogiena** [comment]
yeah, so forget about averages, that's bogus imho

**Comment by noescape00** [comment]
>yeah, so forget about averages, that's bogus imho

we can't correctly estimate how much memory one block takes on average basing on just 1 block. We will end up with really incorrect estimation.

**Comment by dangershony** [comment]
@noescape00  I did try to run your example yeah not pure LRU but good enough.
We just need a guarantee the items within the window time frame are in cache.

I would say block count will fit our needs (max items in cache) instead of calculating the size of each block (even though we could potentially pass that in the signaller.

**Comment by Aprogiena** [comment]
@noescape00 we could get size of everything except for mined block easily without estimation, but @dangershony is probably right that we should start with max items in cache, which i wanted to avoided, but it seems it is quite complicated to achieve that

so the user won't be able to manage this at this moment i guess

ok, let's start with the simplest thing first

**Comment by dangershony** [comment]
Yeah start with the simplest thing so we have a good fix for now, later we can implement our own for optimisation.

**Comment by noescape00** [comment]
>@noescape00 we could get size of everything except for mined block easily without estimation

Not really. You are forgetting the cases where cache is used right now: we ask cache for the block and it doesn't find it, so it pulls it from the repo and adds to the cache. Right now this is the ONLY way we add blocks to the cache and here we need estimation or serialization. 

>ok, let's start with the simplest thing first

>I would say block count will fit our needs (max items in cache) instead of calculating the size of each block 

So I'm going with items limit and not the size limit? 
Ok, perfect.

**Comment by Aprogiena** [comment]
yeah, use items limit, i don't like it, but it is not our main task now, so let's just do it like this and later we can improve it 


> You are forgetting the cases where cache is used right now: we ask cache for the block and it doesn't find it, so it pulls it from the repo and adds to the cache.

you are still very much underestimating me :) i'm not forgetting those, when you deserialize from disk, you have size info, it is just buried very much in the system and it is hard, so that's why i say it is ok to go with items count now. but it is very much possible, just would need rewrite deserialization

**Comment by Aprogiena** [comment]
actually we could do it without rewriting the serialization code, just to save the size with the block to the database. would need resync, but ... anyway, still we start with just items limit
